server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/ShardAssignment.java

```diff
@@ -15,13 +15,18 @@
 
 import static java.util.stream.Collectors.toUnmodifiableSet;
 
-public record ShardAssignment(Set<String> nodeIds, int unassigned, int ignored) {
+public record ShardAssignment(Set<String> nodeIds, int total, int unassigned, int ignored) {
 
-    public boolean isIgnored() {
-        return ignored > 0 && nodeIds.isEmpty();
+    public ShardAssignment {
+        assert total > 0 : "Shard assignment should not be empty";
+        assert nodeIds.size() + unassigned == total : "Shard assignment should account for all shards";
```

Comment:
Can we also assert that `ignored` is one of `0`, `total-1` or `total` depending (respectively) on whether we're ignoring nothing, just replicas, or all copies?